src/common/ult_stack_mmap.c

```diff
@@ -0,0 +1,612 @@
+
+static int
+thread_create_common(struct thread_args *args)
+{
+	const bool is_unnamed              = (args->ta_newthread == NULL);
+	void (*thread_func)(void *)        = args->ta_thread_func;
+	void                   *thread_arg = args->ta_thread_arg;
+	ABT_thread_attr         attr       = ABT_THREAD_ATTR_NULL;
+	size_t                  stack_size = g_smp.smp_thread_stack_size;
+	struct stack_mmap_desc *desc       = NULL;
+	int                     rc;
+
+	if (args->ta_thread_attr != ABT_THREAD_ATTR_NULL) {
+		void *stack = NULL;
+
+		rc = ABT_thread_attr_get_stack(args->ta_thread_attr, &stack, &stack_size);
+		if (unlikely(rc != ABT_SUCCESS)) {
+			D_ERROR("Failed to retrieve ULT stack attributes: " AF_RC "\n", AP_RC(rc));
+			D_GOTO(out, rc);
+		}
+		if (stack != NULL) {
+			D_INFO("Not using mmap stack ULT: using dedicated stack allocator.\n");
+			rc = abt_thread_create_common(args);
+			D_GOTO(out, rc);
+		}
+	}
+
+	/** FIXME migrable and callback properties of the thread attribute will not be duplicated as
+	 * it is not possible to get info on it */
+	rc = ABT_thread_attr_create(&attr);
+	if (unlikely(rc != ABT_SUCCESS)) {
+		D_ERROR("Failed to create ABT thread attr: " AF_RC "\n", AP_RC(rc));
+		D_GOTO(out, rc);
+	}
+
+	rc = smd_find_insert_desc(stack_size, &desc);
+	if (unlikely(rc != 0)) {
+		DL_ERROR(rc, "Not using mmap stack ULT: Failed to find/create stack of size %zu",
+			 stack_size);
+		D_GOTO(error_attr, rc);
+	}
+	D_ASSERT(desc != NULL);
+
+	desc->smd_thread_func = args->ta_thread_func;
+	desc->smd_thread_arg  = args->ta_thread_arg;
+	desc->smd_thread_attr = attr;
+	rc                    = ABT_thread_attr_set_stack(attr, desc->smd_thread_stack, stack_size);
+	if (unlikely(rc != ABT_SUCCESS)) {
+		D_ERROR("Failed to set stack thread attributes : " AF_RC "\n", AP_RC(rc));
+		D_GOTO(error_desc, rc);
+	}
+
+	if (is_unnamed) {
+		thread_func = ult_unnamed_wrapper;
+		thread_arg  = desc;
+	}
+
+	switch (args->ta_flags) {
+	case TCF_ON_POOL:
+		rc = ABT_thread_create(args->ta_pool, thread_func, thread_arg, attr,
+				       args->ta_newthread);
+		break;
+	case TCF_ON_XSTREAM:
+		rc = ABT_thread_create_on_xstream(args->ta_xstream, thread_func, thread_arg, attr,
+						  args->ta_newthread);
+		break;
+	default:
+		D_ERROR("Not using mmap stack ULT: "
+			"Unsupported type of thread creation (type=0x%x)\n",
+			args->ta_flags);
+		D_ASSERT(false);
+		break;
+	}
+	if (unlikely(rc != ABT_SUCCESS)) {
+		D_ERROR("Failed to create ULT : " AF_RC "\n", AP_RC(rc));
+		D_GOTO(error_desc, rc);
+	}
+
+	if (!is_unnamed) {
+		rc = ABT_thread_set_specific(*args->ta_newthread, g_smp.smp_free_mmap_key, desc);
+		if (unlikely(rc != ABT_SUCCESS)) {
+			D_ERROR("Failed to set ULT stack free callback: " AF_RC "\n", AP_RC(rc));
+			D_GOTO(error_thread, rc);
+		}
+	}
+
+	D_DEBUG(DB_MEM, "Created new %s ULT with mmap'ed() stack %p (stack size=%zu)\n",
+		is_unnamed ? "unnamed" : "named", desc->smd_thread_stack, stack_size);
+	D_GOTO(out, rc = ABT_SUCCESS);
```

Comment:
'attr' is leaked.

Reply:
In case of success, `attr` will be freed when the resources of the new created ULT will be deallocated: call to the function ABT_thread_free() for a named thread and automatically at the end of an unnamed ULT.  Thanks to the ABT_key `smp_free_mmap_key`, `attr` will be freed by the `free_mmap_cb()` callback which will be called when the ULT resources will be deallocated.

Reply:
I see, I overlooked that free_mmap_cb() does 'attr' free.

Will that cause issue on error cleanup code path? What if a named thread is created successfully, but the following ABT_thread_set_specific() failed, then on error cleanup phase, the free_map_cb() will be called twice? It looks to me that it's simpler & less error-prone to free the 'attr' in this function directly.

Reply:
In the cleanup code, the attr is freed when needed.  Indeed, if the call to `ABT_thread_set_specific()` fail, then `attr` will only be deleted by the call to `free_mmap_cb(desc)` at line 472 as the function `ABT_thread_free()` should never be called.

However, I agree with you that it is useless to free attr in `free_mmap_cb()` as call to `ABT_thread_create()` is copying its `attr` parameter.  Thus, I am going to delete it asap: just after the call to `ABT_thread_create()`.